README.md
Jitsi Connection Optimization
============
A server side component node.js application that listens for http requests, creates bosh connection on every request and returns rid, sid and jid in the response. That way the client can receive the connection details with a single AJAX request. Than the client can call Strophe.attach with the received rid, sid and jid and start using that connection. This mechanism can speed up the connecting process.

## Installation
TBD

## Building the sources
```
npm install
```

## Running
```
node index.js
```

## Deployment
```
sudo su
curl -sL https://deb.nodesource.com/setup_4.x | sudo -E bash -
apt-get install -y nodejs
wget -qO - https://repo.jitsi.net/debian/unstable/archive.key | sudo apt-key add -
echo 'deb http://download.jitsi.org/nightly/deb unstable/' > /etc/apt/sources.list.d/jitsi-backup.list
apt-get update
apt-get install jiconop
service jiconop start
```

Make sure you meet the following steps: 
- you do not have /etc/nginx/sites-enabled/default (delete it)
- nginx config:
```
#server {
#    listen 80;
#    server_name <your-domain>;
#    return 301 https://$host$request_uri;
#}
server {
    listen 443 ssl default_server;
    listen 80 default_server;
    #server_name <your-domain>;
........
.....
    # BOSH
    location /http-bind {
        proxy_pass      http://localhost:5280/http-bind;
        proxy_set_header X-Forwarded-For $remote_addr;
        proxy_set_header Host beta.meet.jit.si;
    }
}
```
- config.js add property externalConnectUrl:"//<your-domain>/http-pre-bind"

debian/changelog
jiconop (1.0.1-1) unstable; urgency=low

  * Initial release.

 -- Damian Minkov <damencho@jitsi.org>  Mon, 28 Mar 2016 10:30:00 +0200


index.js
var webservice = require('./lib/webservice');

var argv = require('yargs')
    .usage('Usage: $0 [options]')
    .demand('b')
    .alias('b', 'bosh_url')
    .nargs('b', 1)
    .describe('b', 'Bosh url that is going to be used for creating new connections')
    .demand('x')
    .alias('x', 'xmpp_domain')
    .nargs('x', 1)
    .describe('x', 'XMPP Server domain that is going to be used for anonymous login')
    .alias('p', 'port')
    .nargs('p', 1)
    .number('p')
    .describe('p', 'The application will listen for http requests on this port. Default: 9615')
    .help('h')
    .alias('h', 'help')
    .example('$0 -b http://example.jitsi.net/http-bind/ -x example.jitsi.net')
    .argv;

var config = {
    boshUrl: argv.bosh_url,
    XMPPDomain: argv.xmpp_domain,
    port: argv.port || 9615
}

// Starts the webservice that will create the XMPP connection and will return
// rid, sid and jid in the http response.
webservice(config);


lib/XMPPUtils.js
var XMPP = require('node-xmpp-client');

/**
 * Array with the query params from the request that we are going to pass when
 * creating the connection.
 */
var BOSH_QUERY_PARAMS = ["token", "room"];

/**
 * Adds to the given bosh url query parameters that match the names and the
 * prefix from BOSH_QUERY_PARAMS
 * @param bosh the bosh url
 * @param queryParams map with query params that are going to be filtered.
 * @returns the new bosh url
 */
function generateBoshURLWithParams(bosh, queryParams) {
    var boshUrl = bosh;
    for(var key in queryParams) {
        if(!key.length || BOSH_QUERY_PARAMS.indexOf(key) === -1)
            continue;

        boshUrl += (boshUrl.indexOf('?') === -1 ? '?' : '&') + key + "=" +
            queryParams[key];
    }
    return boshUrl;
};

/**
 * Creates xmpp connection and extracts from it rid and sid from the bosh
 * session and jid from the xmpp connection. Returns Promise that is resolved
 * when the connection is done with the extracted information.
 * @param queryParams {object} map with query params that are going to be
 * filtered and passed to the bosh url.
 * @param config {object} - boshUrl, XMPPDomain, port
 */
function createXMPPConnection(queryParams, config) {
    var bosh = generateBoshURLWithParams(config.boshUrl, queryParams);
    return new Promise(function(resolve, reject)
    {
        var connectionInfo = {};
        var prebind = new XMPP({
            jid: "@" + config.XMPPDomain,
            preferred: "ANONYMOUS",
            bosh: {
                url: bosh,
                wait: '60',
                prebind: function(error, data) {
                    if (error){
                        console.log(error);
                        reject(error);
                        return;
                    }
                    connectionInfo = data;
                    prebind.connect();
                }
            }
        });
        prebind.on('online', function() {
            var jid = prebind.jid;
            connectionInfo.jid = jid.local + "@" +
                jid.domain + "/" + jid.resource;
            resolve(connectionInfo);
        });
        prebind.on("error", function (error) {
            console.log(error);
            reject(error);
        })
    });
}

module.exports = {
    createConnection: createXMPPConnection
}


lib/webservice.js
var http = require('http');
var XMPPUtils = require("./XMPPUtils");
var Utils = require("./Utils");

/**
 * Starts web server. On every request tries to create XMPP connection and puts
 * in the response - rid and sid from the bosh session and the jid from the
 * XMPP connection.
 * @param config {object} - boshUrl, XMPPDomain, port
 */
module.exports = function(config) {
    http.createServer(function (req, res) {
        var params = Utils.getQueryParams(req.url);
        XMPPUtils.createConnection(params, config).then(
        function (connectionInfo) {
            res.writeHead(200, {'Content-Type': 'application/json'});
            res.end(JSON.stringify(connectionInfo));
        }).catch(function (error) {
            console.log("Error occured! " + error);
            res.writeHead(500, "Internal Server Error");
            res.end();
        });
    }).listen(config.port);
}


lib/htmlbuilder.js
/**
 * NOTE: That file is not used currently.
 *
 * It's almost the same as webservice.js but in addition instead of returning
 * rid, sid and jid, it adds them to the index.html file and returns it.
 */
var http = require('http');
var fs = require('fs');
var config = require("../config.json");
var index = fs.readFileSync(config.indexHTMLLocation);
var XMPPUtils = require("./XMPPUtils");
var Utils = require("./Utils");

/**
 * Puts passed connectionInfo object in to the passed html.
 * @param html html string
 * @param connectionInfo {object}
 */
function addCredentialsToHML(html, connectionInfo) {
    return html.replace( "</body>","<script>window.XMPPAttachInfoStr = '" +
        JSON.stringify(connectionInfo) + "';</script></body>");
}

/**
 * Starts web server. On every request tries to create XMPP connection and
 * responds with the content of the index.html file with included information
 * about rid and sid from the bosh session and the jid from the
 * XMPP connection.
 */
module.exports = function() {
    http.createServer(function (req, res) {
        var params = Utils.getQueryParams(req.url);
        XMPPUtils.createConnection(params).then(function (connectionInfo) {
            res.writeHead(200, {'Content-Type': 'text/html'});
            res.end(addCredentialsToHML(index.toString(), connectionInfo));
        }).catch(function (error) {
            console.log("Error occured! Fallback to original index.html "
                + error);
            res.writeHead(500, "Internal Server Error");
            res.end(index);
        });
    }).listen(config.port);
};
